File: hs/src/Day8.hs
{-|
Module:         Day8
Description:    <https://adventofcode.com/2021/day/8 Day 8: Seven Segment Search>
-}
{-# LANGUAGE OverloadedStrings #-}
module Day8 (day8a, day8b) where

import Control.Arrow (first)
import Control.Monad (guard)
import Data.Bits ((.&.), (.|.), complement, popCount, setBit)
import Data.Char (ord)
import qualified Data.IntMap as IntMap ((!?), fromListWith)
import Data.List (elemIndex, foldl', partition)
import Data.Text (Text)
import qualified Data.Text as T (breakOnEnd, foldl', length, lines, stripSuffix, words)

day8a :: Text -> Int
day8a input = length $ do
    word <- T.lines input >>= T.words . snd . T.breakOnEnd " | "
    guard $ T.length word `elem` [2, 4, 3, 7]

day8b :: Text -> Maybe Int
day8b input = sum <$> mapM handle (T.lines input)

handle :: Text -> Maybe Int
handle line
  | (Just lhs, rhs) <- first (T.stripSuffix " | ") $ T.breakOnEnd " | " line = do
    let signals = bits <$> T.words lhs
        outputs = bits <$> T.words rhs
        counts = IntMap.fromListWith (<>) [(popCount s, [s]) | s <- signals]
    [one] <- counts IntMap.!? 2
    [seven] <- counts IntMap.!? 3
    [four] <- counts IntMap.!? 4
    ([two], threeFive) <- partition ((== 2) . popCountWithout (four .|. seven)) <$> counts IntMap.!? 5
    ([three], [five]) <- pure $ partition ((== 1) . popCountWithout two) threeFive
    ([six], zeroNine) <- partition ((/= 0) . (one .&.) . complement) <$> counts IntMap.!? 6
    ([zero], [nine]) <- pure $ partition ((== 2) . popCountWithout three) zeroNine
    [eight] <- counts IntMap.!? 7
    let digits = [zero, one, two, three, four, five, six, seven, eight, nine]
    foldl' (\x y -> 10 * x + y) 0 <$> mapM (`elemIndex` digits) outputs
  | otherwise = Nothing
  where
    bits = T.foldl' (\acc c -> acc `setBit` (ord c .&. 31)) (0 :: Int)
    popCountWithout antimask = popCount . (.&. complement antimask)

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day8.kt
package com.github.ephemient.aoc2021

/** Day 8: Seven Segment Search */
class Day8(private val lines: List<String>) {
    fun part1(): Int = lines.sumOf {
        it.substringAfter(" | ")
            .split(' ')
            .count { it.length == 2 || it.length == 4 || it.length == 3 || it.length == 7 }
    }

    fun part2(): Int = lines.sumOf { line ->
        val (lhs, rhs) = line.split(" | ", limit = 2)
        val signals = lhs.split(' ').map { it.toBits() }.groupBy { it.countOneBits() }
        val one = signals.getValue(2).single()
        val seven = signals.getValue(3).single()
        val four = signals.getValue(4).single()
        val fourSeven = four or seven
        val (twos, threeFive) = signals.getValue(5).partition { (it and fourSeven.inv()).countOneBits() == 2 }
        val two = twos.single()
        val (threes, fives) = threeFive.partition { (it and two.inv()).countOneBits() == 1 }
        val three = threes.single()
        val five = fives.single()
        val (sixes, zeroNine) = signals.getValue(6).partition { one and it.inv() != 0 }
        val six = sixes.single()
        val (zeros, nines) = zeroNine.partition { (it and three.inv()).countOneBits() == 2 }
        val nine = nines.single()
        val zero = zeros.single()
        val eight = signals.getValue(7).single()
        val digits = intArrayOf(zero, one, two, three, four, five, six, seven, eight, nine)
        rhs.split(' ').fold(0) { acc: Int, s -> 10 * acc + digits.indexOf(s.toBits()) }
    }

    companion object {
        private fun String.toBits(): Int = fold(0) { acc, c -> acc or 1.shl(c.code) }
    }
}
